transport_linux.go

```diff
@@ -113,10 +116,20 @@ func (c *conn) timeSinceLastActive() time.Duration {
 
 func (c *conn) Close() error {
 	select {
-	case <-c.close:
-		return nil
+	case <-c.closed:
+		// already closed
 	default:
-		close(c.close)
-		return nil
+		// Ask server to close this conn so that it happens in the dispatch loop
+		c.s.requestCloseConn(c)
+	}
+	return nil
+}
+
+func (c *conn) doClose() {
+	select {
+	case <-c.closed:
+		// already closed
+	default:
+		close(c.closed)
```

Comment:
This still allows for a race condition as two routines could both enter the default case.  One would end up trying to close and already closed channel and panic.  You could use a `sync.Once` instead.

Reply:
Good idea!